src/txrebroadcast.cpp

```diff
@@ -0,0 +1,273 @@
+        const uint256& wtxid = tx->GetWitnessHash();
+
+        // Check if we have kinda recently rebroadcasted this transaction the
+        // maximum number of times.
+        if (m_max_filter.contains(wtxid)) continue;
+
+        // Check if we have previously rebroadcasted, decide if we will this
+        // round, and if so, record the attempt.
+        auto entry_it = m_attempt_tracker->find(wtxid);
+
+        if (entry_it == m_attempt_tracker->end()) {
+            // No existing entry, we will rebroadcast, so create a new one
+            RebroadcastEntry entry(start_time, wtxid);
+            m_attempt_tracker->insert(entry);
+        } else if (entry_it->m_count >= MAX_REBROADCAST_COUNT) {
+            // We have already rebroadcast this transaction the maximum number
+            // of times permitted. Record in the max filter, remove from
+            // attempt tracker, and skip rebroadcasting.
+            m_max_filter.insert(wtxid);
+            const auto it = m_attempt_tracker->find(tx->GetWitnessHash());
+            if (it != m_attempt_tracker->end()) m_attempt_tracker->erase(it);
+            continue;
+        } else if (entry_it->m_last_attempt > start_time - MIN_REATTEMPT_INTERVAL) {
+            // We already rebroadcasted this in the past 4 hours. Even if we
+            // added it to the set, it would probably not get INVed to most
+            // peers due to filterInventoryKnown.
+            continue;
+        } else {
+            // We have rebroadcasted this transaction before, but will try
+            // again now. Record the attempt.
+            auto UpdateRebroadcastEntry = [start_time](RebroadcastEntry& rebroadcast_entry) {
+                rebroadcast_entry.m_last_attempt = start_time;
+                ++rebroadcast_entry.m_count;
+            };
+
+            m_attempt_tracker->modify(entry_it, UpdateRebroadcastEntry);
+        }
+
+        // Add to set of rebroadcast candidates
+        rebroadcast_txs.push_back(TxIds(txid, wtxid));
+    }
+
+    TrimMaxRebroadcast();
+
+    auto delta1 = after_cnb_time - start_time;
+    auto delta2 = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "GetRebroadcastTransactions(): %d us total, %d us spent in CreateNewBlock.\n", delta2.count(), delta1.count());
+    LogPrint(BCLog::NET, "Queued %d transactions for attempted rebroadcast, filtered from %d candidates with cached fee rate of %s.\n", rebroadcast_txs.size(), block_template->block.vtx.size() - 1, m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB));
+
+    for (TxIds ids : rebroadcast_txs) {
+        LogPrint(BCLog::NET, "Attempting to rebroadcast txid: %s, wtxid: %s\n", ids.m_txid.ToString(), ids.m_wtxid.ToString());
+    }
+
+    return rebroadcast_txs;
+};
+
+void TxRebroadcastHandler::CacheMinRebroadcastFee()
+{
+    CChainState& chain_state = m_chainman.ActiveChainstate();
+    if (chain_state.IsInitialBlockDownload()) return;
+    auto tip = m_chainman.ActiveTip();
```

Comment:
I think you should be locking `cs_main` prior to calling this -- you want a consistent view between `tip` and `current_fee_rate`, but currently there's no reason you couldn't update the tip immediately after this line, thus associating the old tip with the new (much reduced) fee rate.

Reply:
yeah makes sense, done